### PR TITLE
Feat/intern 1 create booking

### LINKED ISSUE #22: Seeder master data

### LINKED ISSUE #24: Add Booking

### PR DESCRIPTION
**Pull request by hnthinh55**

### CODE REVIEW

File: TicketBooking.Data/Infrastructure/GenericRepository.cs [hunk 1]
```diff
@@ -53,7 +53,7 @@ public async Task<IEnumerable<T>> GetAll(params string[] includes)
             return query;
         }
 
-        public async Task<T> GetById(Guid id, params string[] includes)
+        public async Task<T> GetById(Guid? id, params string[] includes)
         {
```
Comment (tanngocIT) [inline TicketBooking.Data/Infrastructure/GenericRepository.cs]: I don't think Id can be null.

File: TicketBooking.Data/Infrastructure/GenericRepository.cs [hunk 2]
```diff
@@ -63,7 +63,7 @@ public async Task<T> GetById(Guid id, params string[] includes)
             return model;
         }
 
-        public async Task<bool> Remove(Guid id)
+        public async Task<bool> Remove(Guid? id)
         {
```
Comment (tanngocIT) [inline TicketBooking.Data/Infrastructure/GenericRepository.cs]: I don't think Id can be null.

File: TicketBooking.Data/Infrastructure/GenericRepository.cs [hunk 3]
```diff
@@ -76,13 +76,12 @@ public async Task<bool> Remove(Guid id)
...
+        public void Update(T entity)
         {
-            this._context.Entry<T>(entity).State = EntityState.Modified;
-            return true;
+            dbSet.Update(entity);
         }
 
-        public async Task<T> GetById(Guid id)
+        public async Task<T> GetById(Guid? id)
         {
```
Comment (tanngocIT) [inline TicketBooking.Data/Infrastructure/GenericRepository.cs]: I don't think Id can be null.

File: TicketBooking.Service/Services/BookingListService/BookingListService.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+            this.bookingListRepo = bookingListRepo;
+            this.unitOfWork = unitOfWork;
+            this.mapper = mapper;
+        }
+
+        public async Task<bool> AddBookingList(BookingListViewModel bookingList)
+        {
+            var contactDetail = mapper.Map<BookingList>(bookingList);
+            await bookingListRepo.Add(contactDetail);
+            await unitOfWork.CompletedAsync();
```
Comment (tanngocIT) [inline TicketBooking.Service/Services/BookingListService/BookingListService.cs:30]: unit of work should return true/false

File: TicketBooking.Service/Services/BookingService/BookingService.cs [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+                    };
+                }
+                return new Response
+                {
+                    Status = true,
+                    Message = "Invalid flight ",
+                    Data = null
+                };
+            }
+            catch (InvalidOperationException)
```
Comment (tanngocIT) [inline TicketBooking.Service/Services/BookingService/BookingService.cs]: no need try catch

File: TicketBooking.Service/Services/BookingService/BookingService.cs [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+                    NumberPeople = model.NumberPeople,
+                    DateBooking = DateTime.Now,
+                    Reference = model.Reference,
+                    TotalPrice = 0,
+                    IsPaid = model.IsPaid,
+                    IsRoundFlight = model.IsRoundFlight,
+                    Status = model.Status,
+
+                };
+                await bookingRepo.AddBooking(booking);
```
Comment (tanngocIT) [inline TicketBooking.Service/Services/BookingService/BookingService.cs]: should validate all info before adding them to DB

File: TicketBooking.Service/Services/BookingService/BookingService.cs [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+                {
+
+                    if (booking.IsRoundFlight != true)
+                    {
+                        var bookingListVM = new BookingListViewModel
+                        {
+                            NumberSeat = booking.NumberPeople,
+                            BookingId = booking.Id,
+                            FlightId = model.FlightId,
+                            FlightPrice = 0
```
Comment (tanngocIT) [inline TicketBooking.Service/Services/BookingService/BookingService.cs]: price = 0 ?

File: TicketBooking.Service/Services/BookingService/BookingService.cs [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+            }
+
+        }
+
+        public async Task<Response> CancelBooking(Guid bookingId)
+        {
+            var booking = await bookingRepo.GetById(bookingId);
+            if (booking != null)
+            {
+                booking.Status = "Cancel";
```
Comment (tanngocIT) [inline TicketBooking.Service/Services/BookingService/BookingService.cs]: booking status should have enum or constant

File: TicketBooking.Service/Services/BookingService/BookingService.cs [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+            bookingListRepo = bookingList;
+            bookingSeatRepo = bookingSeat;
+            passengerRepo = passenger;
+            this.contactRepo = contactRepo;
+            this.flightRepo = flightRepo;
+            bookingServiceRepo = serviceRepository;
+            extraServiceRepo = extra;
+        }
+
+        public async Task<Response> RequestBooking(BookingRequestModel model)
```
Comment (tanngocIT) [inline TicketBooking.Service/Services/BookingService/BookingService.cs]: need to  have ExtraService for this

File: TicketBooking.Service/Services/BookingService/BookingService.cs (hunk 7 above)
Comment (tanngocIT) [reply TicketBooking.Service/Services/BookingService/BookingService.cs]: Missing type of Seat

File: TicketBooking.Service/Services/BookingService/BookingService.cs [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+                Message = "Cancel failed!! Invalid booking"
+
+            };
+        }
+
+        public async Task<Response> AddBookingService(List<Guid> extraServices, Guid bookingList)
+        {
+            var checkBooking = await bookingListRepo.GetById(bookingList);
+            decimal sum = 0;
+            if (checkBooking != null)
```
Comment (tanngocIT) [inline TicketBooking.Service/Services/BookingService/BookingService.cs]: `
if(checkBooking == null)
   return ......

if (extraServices.Count <= 0)
   return ......
`

File: TicketBookingAPI/Controllers/BookingController.cs [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+            if (ModelState.IsValid)
+            {
+                try
+                {
+                    logger.LogInformation("Start Booking");
+                    var result = await bookingService.RequestBooking(model);
+                    return StatusCode(StatusCodes.Status201Created, Ok(result));
+                }
+                catch (Exception e)
+                {
```
Comment (tanngocIT) [inline TicketBookingAPI/Controllers/BookingController.cs]: shouldn't try/catch

File: TicketBookingAPI/Controllers/BookingController.cs [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        {
+            if (ModelState.IsValid)
+            {
+                try
+                {
+                    logger.LogInformation("Add more service");
+                    var result = await bookingService.AddBookingService(extraServices, bookingList);
+                    return StatusCode(StatusCodes.Status201Created, Ok(result));
+                }
+                catch (Exception e)
```
Comment (tanngocIT) [inline TicketBookingAPI/Controllers/BookingController.cs]: shouldn't try/catch

File: TicketBookingAPI/Controllers/BookingController.cs [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+        {
+            if (ModelState.IsValid)
+            {
+                try
+                {
+                    logger.LogInformation("Add more service");
+                    var result = await bookingService.CancelBooking(bookingId);
+                    return StatusCode(StatusCodes.Status201Created, Ok(result));
+                }
+                catch (Exception e)
```
Comment (tanngocIT) [inline TicketBookingAPI/Controllers/BookingController.cs]: shouldn't try/catch

File: TicketBookingAPI/Controllers/BookingController.cs [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+                {
+                    logger.LogError("Add service failed");
+                    return BadRequest(e.Message);
+                }
+            }
+            return BadRequest(ModelState);
+        }
+
+        [HttpPost("request-contact")]
+        public async Task<IActionResult> CreateContact(ContactViewModel contact)
```
Comment (tanngocIT) [inline TicketBookingAPI/Controllers/BookingController.cs]: [FromBody]

File: TicketBookingAPI/Controllers/BookingController.cs [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+        public BookingController(IBookingService bookingService, IContactDetailServcie contactService, ILogger<BookingController> logger)
+        {
+            this.bookingService = bookingService;
+            this.contactService = contactService;
+            this.logger = logger;
+        }
+
+        [HttpPost("request-booking")]
+        public async Task<IActionResult> RequestBooking(BookingRequestModel model)
```
Comment (tanngocIT) [inline TicketBookingAPI/Controllers/BookingController.cs]: [FromBody]

Comment (hnthinh55) [reply TicketBookingAPI/Controllers/BookingController.cs]: fixed

File: TicketBookingAPI/Controllers/BookingController.cs (hunk 14 above)
Comment (hnthinh55) [reply TicketBookingAPI/Controllers/BookingController.cs]: fixed

File: TicketBookingAPI/Controllers/BookingController.cs (hunk 13 above)
Comment (hnthinh55) [reply TicketBookingAPI/Controllers/BookingController.cs]: fixed

File: TicketBookingAPI/Controllers/BookingController.cs (hunk 12 above)
Comment (hnthinh55) [reply TicketBookingAPI/Controllers/BookingController.cs]: fixed

File: TicketBookingAPI/Controllers/BookingController.cs (hunk 11 above)
Comment (hnthinh55) [reply TicketBookingAPI/Controllers/BookingController.cs]: fixed

File: TicketBooking.Service/Services/BookingService/BookingService.cs (hunk 5 above)
Comment (hnthinh55) [reply TicketBooking.Service/Services/BookingService/BookingService.cs]: fix

File: TicketBooking.Data/Infrastructure/GenericRepository.cs (hunk 3 above)
Comment (hnthinh55) [reply TicketBooking.Data/Infrastructure/GenericRepository.cs]: fix

File: TicketBooking.Data/Infrastructure/GenericRepository.cs (hunk 2 above)
Comment (hnthinh55) [reply TicketBooking.Data/Infrastructure/GenericRepository.cs]: fixed

File: TicketBooking.Data/Infrastructure/GenericRepository.cs (hunk 1 above)
Comment (hnthinh55) [reply TicketBooking.Data/Infrastructure/GenericRepository.cs]: fixed

File: TicketBooking.Service/Services/BookingService/BookingService.cs (hunk 8 above)
Comment (hnthinh55) [reply TicketBooking.Service/Services/BookingService/BookingService.cs]: fixed

File: TicketBooking.Service/Services/BookingListService/BookingListService.cs (hunk 4 above)
Comment (hnthinh55) [reply TicketBooking.Service/Services/BookingListService/BookingListService.cs:30]: fixed

**Review by tanngocIT — APPROVED** [review]

**Pull request merged**